File: packages_clj/clarity/src/io/relica/clarity/services/entity_model_service.clj
(ns io.relica.clarity.services.entity-model-service
  (:require [clojure.tools.logging :as log]
            [clojure.pprint :refer [pprint]]
            [clojure.core.async :refer [go <!]]
            [clojure.spec.alpha :as s]
            [io.relica.clarity.io.archivist-api :as archivist-api]))

;; ------------------------------------------------------------------ HELPERS --

;; (defn retrieve-supertypes-and-definitions [uid]
;;   "Retrieve supertypes and definitions for a physical object entity"
;;   (go
;;     (try
;;       (let [definitive-facts (<! (archivist-api/get-definitive-facts uid))
;;             names (map :lh_object_name definitive-facts)
;;             supertypes (map :rh_object_uid definitive-facts)
;;             definitions (map :full_definition definitive-facts)]
;;         {:names names
;;          :supertypes supertypes
;;          :definitions definitions})
;;       (catch Exception e
;;         (log/error e "Failed to retrieve supertypes and definitions")
;;         {:valid false :error (str "Error: " (.getMessage e)) :uid uid}))))

(defn retrieve-classifiers [uid]
  "Retrieve classifiers for an individual entity"
  (go
    (try
      (let [classifiers (<! (archivist-api/get-definitive-facts uid))]
        (log/info "Classifiers: " classifiers)
        classifiers)
      (catch Exception e
        (log/error e "Failed to retrieve classifiers")
        {:valid false :error (str "Error: " (.getMessage e)) :uid uid}))))

(defn retrieve-stuff [uid]
  "Retrieve stuff for an individual or kind of entity"
  (go
    (try
      (let [stuff (<! (archivist-api/get-all-related-facts uid))
            classifications (filter #(= (:rel_type_uid %) 1225) stuff)
            specializations (filter #(= (:rel_type_uid %) 1146) stuff)
            qualifications (filter #(= (:rel_type_uid %) 1726) stuff)
            synonyms (vec
                      (reduce (fn [m f]
                                (conj m (:lh_object_name f)))
                                  #{}
                                  (filter #(= (:rel_type_uid %) 1981) stuff)
                                  ))]
        (log/info "Stuff: " stuff)
        {:classifications classifications
         :specializations specializations
         :qualifications qualifications
         :synonyms synonyms})
      (catch Exception e
        (log/error e "Failed to retrieve stuff")
        {:valid false :error (str "Error: " (.getMessage e)) :uid uid}))))

;; --------------------------------------------------------------------- KIND --

(defn retrieve-kind-of-entity-model
  "Retrieve and transform an entity object to its semantic model representation"
  [uid]
  (go
    (try
      (let [definitive-facts (<! (archivist-api/get-definitive-facts uid))
            names (map :lh_object_name definitive-facts)
            supertypes (map :rh_object_uid definitive-facts)
            definitions (map :full_definition definitive-facts)
            stuff (<! (archivist-api/get-all-related-facts uid))
            synonym-facts (filter #(= (:rel_type_uid %) 1981) stuff)
            synonym-names (map :lh_object_name synonym-facts)
            facts (vec (concat
                   definitive-facts
                   synonym-facts))]
        (log/info "Supertypes:" (count supertypes))
        (log/info "Definitions:" (count definitions))
        (log/info "Names:" (count names) "First name:" (first names))
        (log/info "Synonyms:" synonym-names)
        ;; {:uid uid
        ;;  :name (first names)
        ;;  :nature :kind
        ;;  :definitions definitions
        ;;  :supertypes supertypes
        ;;  :classifications (:classifications stuff)
        ;;  :specializations (:specializations stuff)
        ;;  :qualifications (:qualifications stuff)
        ;;  :synonyms (:synonyms stuff)}

        {:uid uid
         :name (first names)
         :nature :kind
         :definitions definitions
         :supertypes supertypes
         :synonyms synonym-names
         :facts facts})
      (catch Exception e
        (log/error e "Failed to retrieve entity object model")
        {:valid false :error (str "Error: " (.getMessage e)) :uid uid}))))

;; --------------------------------------------------------------- IDNIVIDUAL --

(defn retrieve-individual-entity-model
  "Retrieve and transform an entity object to its semantic model representation"
  [uid]
  (go
    (try
      (let [classifiers (<! (retrieve-classifiers uid))]
        (log/info "Classifiers:" (count classifiers))
        {:uid uid
         :name (:lh_object_name (first classifiers))
         :nature :individual
         :classifiers (map #(:rh_object_uid %) classifiers)
         :facts classifiers})
      (catch Exception e
        (log/error e "Failed to retrieve entity individual model")
        {:valid false :error (str "Error: " (.getMessage e)) :uid uid}))))

File: packages/clarity/src/services/entity-model.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ArchivistService } from '../archivist/archivist.service';

export interface EntityModelData {
  names: string[];
  supertypes: number[];
  definitions: string[];
}

export interface EntityStuffData {
  classifications: any[];
  specializations: any[];
  qualifications: any[];
  synonyms: string[];
}

export interface KindEntityModel {
  uid: number;
  name: string;
  nature: 'kind';
  definitions: string[];
  supertypes: number[];
  synonyms: string[];
}

export interface IndividualEntityModel {
  uid: number;
  name: string;
  nature: 'individual';
  classifiers: number[];
}

@Injectable()
export class EntityModelService {
  private readonly logger = new Logger(EntityModelService.name);

  constructor(private readonly archivistService: ArchivistService) {}

  /**
   * Retrieve supertypes and definitions for an entity
   */
  async retrieveSupertypesAndDefinitions(
    uid: number,
  ): Promise<EntityModelData> {
    try {
      const definitiveFacts =
        await this.archivistService.getDefinitiveFacts(uid);
      const names = definitiveFacts.map((f) => f.lh_object_name);
      const supertypes = definitiveFacts.map((f) => f.rh_object_uid);
      const definitions = definitiveFacts.map((f) => f.full_definition);

      return {
        names,
        supertypes,
        definitions,
      };
    } catch (error) {
      this.logger.error(
        `Failed to retrieve supertypes and definitions for UID ${uid}:`,
        error,
      );
      throw new Error(`Error retrieving entity data: ${error.message}`);
    }
  }

  /**
   * Retrieve classifiers for an individual entity
   */
  async retrieveClassifiers(uid: number) {
    try {
      const classifiers = await this.archivistService.getDefinitiveFacts(uid);
      this.logger.debug(
        `Retrieved ${classifiers.length} classifiers for UID ${uid}`,
      );
      return classifiers;
    } catch (error) {
      this.logger.error(
        `Failed to retrieve classifiers for UID ${uid}:`,
        error,
      );
      throw new Error(`Error retrieving classifiers: ${error.message}`);
    }
  }

  /**
   * Retrieve related facts for an entity (classifications, specializations, etc.)
   */
  async retrieveStuff(uid: number): Promise<EntityStuffData> {
    try {
      const stuff = await this.archivistService.retrieveAllFacts(uid);

      // Filter by relation type UIDs (matching Clojure implementation)
      const classifications = stuff.filter((f) => f.rel_type_uid === 1225);
      const specializations = stuff.filter((f) => f.rel_type_uid === 1146);
      const qualifications = stuff.filter((f) => f.rel_type_uid === 1726);

      // Extract unique synonyms
      const synonyms: string[] = Array.from(
        new Set(
          stuff
            .filter((f) => f.rel_type_uid === 1981)
            .map((f) => f.lh_object_name)
            .filter((name): name is string => typeof name === 'string'),
        ),
      );

      this.logger.debug(
        `Retrieved stuff for UID ${uid}: ${classifications.length} classifications, ${specializations.length} specializations, ${qualifications.length} qualifications, ${synonyms.length} synonyms`,
      );

      return {
        classifications,
        specializations,
        qualifications,
        synonyms,
      };
    } catch (error) {
      this.logger.error(`Failed to retrieve stuff for UID ${uid}:`, error);
      throw new Error(`Error retrieving related facts: ${error.message}`);
    }
  }

  /**
   * Retrieve and transform a kind entity to its semantic model representation
   */
  async retrieveKindOfEntityModel(uid: number): Promise<KindEntityModel> {
    try {
      const { names, supertypes, definitions } =
        await this.retrieveSupertypesAndDefinitions(uid);
      const stuff = await this.retrieveStuff(uid);

      this.logger.log(
        `Retrieved kind entity model for UID ${uid}: ${names[0]}`,
      );

      return {
        uid,
        name: names[0] || '',
        nature: 'kind',
        definitions,
        supertypes,
        synonyms: stuff.synonyms,
      };
    } catch (error) {
      this.logger.error(
        `Failed to retrieve kind entity model for UID ${uid}:`,
        error,
      );
      throw new Error(`Error retrieving kind entity model: ${error.message}`);
    }
  }

  /**
   * Retrieve and transform an individual entity to its semantic model representation
   */
  async retrieveIndividualEntityModel(
    uid: number,
  ): Promise<IndividualEntityModel> {
    try {
      const classifiers = await this.retrieveClassifiers(uid);

      this.logger.log(`Retrieved individual entity model for UID ${uid}`);

      return {
        uid,
        name: classifiers[0]?.lh_object_name || '',
        nature: 'individual',
        classifiers: classifiers.map((c) => c.rh_object_uid),
      };
    } catch (error) {
      this.logger.error(
        `Failed to retrieve individual entity model for UID ${uid}:`,
        error,
      );
      throw new Error(
        `Error retrieving individual entity model: ${error.message}`,
      );
    }
  }
}
